crates/red_knot_project/src/metadata.rs

```
@@ -62,22 +65,36 @@ impl ProjectMetadata {
 
     /// Loads a project from a set of options with an optional pyproject-project table.
     pub(crate) fn from_options(
-        options: Options,
+        mut options: Options,
         root: SystemPathBuf,
         project: Option<&Project>,
-    ) -> Self {
+    ) -> Result<Self, TooLargeRequiresPythonError> {
         let name = project
-            .and_then(|project| project.name.as_ref())
-            .map(|name| Name::new(&***name))
+            .and_then(|project| project.name.as_deref())
+            .map(|name| Name::new(&**name))
             .unwrap_or_else(|| Name::new(root.file_name().unwrap_or("root")));
 
-        // TODO(https://github.com/astral-sh/ruff/issues/15491): Respect requires-python
-        Self {
+        // If the project doesn't specify a python version but the `project.requires-python` field is set, resolve the version
+        if let Some(project) = project {
+            if !options
+                .environment
+                .as_ref()
+                .is_some_and(|env| env.python_version.is_some())
+            {
+                if let Some(requires_python) = project.resolve_requires_python_lower_bound()? {
+                    let mut environment = options.environment.unwrap_or_default();
+                    environment.python_version = Some(requires_python);
```

Should we do something like building the minimum between the `requires_python` lower bound and our minimum supported version here?

I find it surprising that I get a Python version of 3.9 if I don't specify anything, but a Python version of 3.6 if I set `requires-python = ">= 3.6"`.

Conversely, it might be surprising that upper bounds are ignored completely. If I set `requires-python = "<3.6"`, I still get a Python version of 3.9.

---

That's a great catch. I'm not sure if it's worth special casing because:

* https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#python-requires Calls it out that it is the minimum python version supported and using `<` doesn't make sense
* https://discuss.python.org/t/requires-python-upper-limits/12663 and later https://discuss.python.org/t/requires-python-upper-limits/12663/84 both mention that the field is intended as lower bound and not an upper bound. Pip doesn't seem to support the upper bound. (Shared by @konstin )

We could error if we see a requires pyhon constraint without a lower bound over silently ignoring. I'd be interested in @AlexWaygood or @carljm take on this.

---

Interesting question.

I wondered whether it might be possible for red-knot to have users who have a pyproject.toml file (and a `requires-python` field in that pyproject.toml file) without actually ever intending their project to be built as a package while a wheel. If so, the packaging.python.org guide might not be relevant here; they might _only_ be using the `requires-python` field as a configuration setting for tools such as Ruff and red-knot. But on second thought, I don't think this is worth worrying about _too_ much. The `requires-python` field is part of the `[project]` table in `pyproject.toml`; it seems reasonable to assume that users who include a `[project]` table are intending their code to be built and distributed as a wheel. So I think we _can_ safely assume "packaging semantics" for the `requires-python` field.

It might be good to print a warning to the terminal if we resolve the user's configuration settings to a `requires-python` version less than typeshed's oldest supported version. Typeshed exposes this information in [this field here](https://github.com/python/typeshed/blob/24c78b9e0dff457275092025a14387dac206f5d6/pyproject.toml#L173-L174) in its pyproject.toml file (the intention is that it could be used by tools vendoring typeshed as well as by typeshed's internal tools). It doesn't look like we currently vendor typeshed's pyproject.toml file as part of our typeshed syncs, but we could easily start doing so. The reason why I think we should print a warning is that people might think that we'll accurately understand the Python 3.7 stdlib if they put `requires-python = ">= 3.7"` in their pyproject.toml file -- but we won't, because typeshed deleted the pre-Python-3.8 branches when it dropped support for Python 3.7. The reason why it should be a suppressible warning rather than an error is that the user might have Python <3.8 branches in their own code.

If we see `requires-python = "<3.6"`, I'd prefer us to error (with a helpful error message telling them to set the red-knot `python-version` setting telling us what to do) rather than fallback to the default of Python 3.9. Falling back to Python 3.9 here feels too much like guesswork.

---

I think there are two separate issue raised in @sharkdp 's comment.

To me the more important issue is, what happens if someone specifies a `requires-python` lower bound that is lower than our minimum supported Python version? I think in this case we have to just check with our minimum supported Python version, but maybe we should also emit a warning (which would be silenced by explicitly setting a red-knot python version), so people don't mistakenly think they are getting red-knot coverage of Python 3.6?

The second issue is, what happens if someone specifies a `requires-python` with no lower bound? I agree that this is not a supported use of `requires-python` generally, but I do also think that means we should emit a warning on it.

---

I created an issue for the first case (by simply citing Alex) because it's unrelated to this PR (it doesn't matter if you use requires python or python-version, the problem applies to both settings). 

I'll go and add an error for the "no lower bound case" because erroring at this stage is easier than emitting a warning and I don't feel like this edge case is worth complicating the architecture to make it a warning.